keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,986 @@
+links can take a lot of time, tens of minutes or more. This often results
+in a bad user experience when deploying a workload triggers pulling such
+an image as a side-effect. For a long time the user has no visibility to
+how creating the workload progresses, or if it progresses at all.
+
+Introducing a new CRI ImageService RPC call for pulling images with a
+streaming progress notification response would provide the low-level
+building blocks to improve this situation.
+
+
+
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Extend CRI to provide image pulling progress API that can be utilized by client
+tools (crictl pull), or machinery (kubelet pulling the image)
+- Implement PoC / draft for CRI PullImageWithProgress call to runtime in Kubelet for API review purposes
+
+### Non-Goals
+
+- implementation of utilization of proposed interface additions by the
+client-tools and / or k8s services
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+We propose introducing new (additional, not replacing old one) API for requesting image pull,
+that will return stream with periodic updates sent through it to the client. The image pull
+request parameters will contain the type of granularty based on which the client wants to receive
+updates about the progress:
+- time-based
+- size-based
+
+and the frequency of the updates:
+- every N seconds
+- every N Kibibytes of the total image size being downloaded
+respectively to the type
+
+If / when it is possible to reliably determine percentage of the progress in the runtime,
+percentage-based granularity type can be introduced then.
+
+If the client did not specify the preferred notification granularity, default values should be used,
+for instance every Gibibyte downloaded.
+
+For kubelet we propose the default granularity to be time-based with 30 seconds interval, so
+Pod object will get an event with image pulling progress every 30 seconds.
+The granularity type and interval should be configurable through kubelet-config.
```

Comment:
can we add more details about the proposed kubelet config change, i.e. config name and type?

Reply:
Done.